connectors/uplynk/src/main/java/com/theoplayer/android/connector/uplynk/internal/UplynkAdIntegration.kt

```diff
@@ -11,23 +11,52 @@ import com.theoplayer.android.connector.uplynk.network.UplynkApi
 internal class UplynkAdIntegration(
     val theoplayerView: THEOplayerView,
     val controller: ServerSideAdIntegrationController,
-    val uplynkDescriptionConverter: UplynkSsaiDescriptionConverter
+    val eventDispatcher: UplynkEventDispatcher,
+    val uplynkDescriptionConverter: UplynkSsaiDescriptionConverter,
+    private val uplynkApi: UplynkApi
 ) : ServerSideAdIntegrationHandler {
 
     private val player: Player
         get() = theoplayerView.player
 
-    private val uplynkApi = UplynkApi()
-
     override suspend fun setSource(source: SourceDescription): SourceDescription {
-        val uplynkSource = source.sources.find { it.ssai is UplynkSsaiDescription } ?: return source
-        val ssaiDescription = uplynkSource.ssai as? UplynkSsaiDescription ?: return source
-        val response = uplynkApi.preplay(uplynkDescriptionConverter.buildPreplayUrl(ssaiDescription))
+
+        val uplynkSource = source.sources.singleOrNull { it.ssai is UplynkSsaiDescription }
+        val ssaiDescription = uplynkSource?.ssai as? UplynkSsaiDescription ?: return source
+
+        val response = uplynkDescriptionConverter
+            .buildPreplayUrl(ssaiDescription)
+            .let { uplynkApi.preplay(it) }
+            .also {
+                try {
+                    eventDispatcher.dispatchPreplayEvents(it.parseExternalResponse())
+                } catch (e: Exception) {
+                    eventDispatcher.dispatchPreplayFailure(e)
+                    controller.error(e)
+                }
+            }
+
+        val internalResponse = response.parseMinimalResponse()
 
         val newSource = source.replaceSources(source.sources.toMutableList().apply {
             remove(uplynkSource)
-            add(0, uplynkSource.replaceSrc(response.playURL))
+            add(0, uplynkSource.replaceSrc(internalResponse.playURL))
```

Comment:
unrelated to this PR, but I think here we miss the [playbackUrlParameters](https://www.theoplayer.com/docs/theoplayer/v6/api-reference/ios/Classes/VerizonMediaSource.html#/s:13THEOplayerSDK18VerizonMediaSourceC21playbackUrlParametersSDyS2SGSgvp) from here that has to be appended to the final (resolved/parsed) playURL.

(maybe it is coming in the next iteration of the feature, just do not forget it :) )